whole-tale/ngx-dashboard
# fix: add Edit Mode back to Run > Metadata, add relatedIdentifiers, fix publishInfo

## Problem
Edit Mode was removed by a recent change. We like this pattern and wish to continue using it in the Metadata view. Without this feature, the user has no way to "Preview" what the rendered view of the Metadata will look like for users who do not have edit access.


## Approach
* Add back the Edit Mode functions, style the Save/Edit/Cancel buttons to look a bit better

NOTE: this does not address the issue of unsaved edits polluting frontend data

Also fix various minor issues with the rendered metadata view:
* Sort `publishInfo` by date to grab latest item
* Add the missing `relatedIdentifiers` section to the rendered metadata (this is not directly editable.. should it be?)
* Add a left-margin to values to space them out ever so slightly from their labels

Fixes #95 

## How to Test
Prerequisites: at least one Tale created that you own, at least one dataset registered

1. Checkout this branch locally, rebuild the dashboard
2. Login to the WholeTale Dashboard
3. Navigate to the Tale Catalog, hover over a Tale that you own, and click View
    * You should be brought to the Run > Metadata view (if the Tale is not running)
4. If necessary, click over to the Run > Metadata view
    * You should see the rendered (not editable) Metadata view is presented
5. Scroll down and click Edit (this is hidden if you cannot edit the Tale)
    * You should now see the editable Metadata form
    * Scroll all the way down and you should see two buttons alongside the "Public?" slider: Save and Cancel
6. Change the title of the Tale or Authors, but don't Save or Cancel yet
    * You should notice the header at the top update as the Tale's metadata changes
7. Scroll back down and click "Cancel"
    * You should see your Tale revert to its state from before you clicked Edit
8. Click Edit again, make some edits, and click Save
    * You should see the rendered Metadata view
    * You should see your Tale's updated metadata reflected in the rendered values


#### `relatedIdentifiers`
1. Attach a dataset to your Tale on Run > Files
2. Navigate to Run > Metadata and check the rendered Metadata view
    * You should see a Related Identifiers section now appears that will cite related DOIs

#### `publishInfo`
1. Publish your Tale a few times to DataONE or Zenodo
2. Check the Run > Metadata view for "Published Location"
    * You should see that the "Published Location" contains a link and a DOI
3. On Girder's API explorer, perform a GET /tale, and find your Tale
    * You should see several entries under the Tale's `publishInfo` field
4. Take note of the `pid` and `date` for your latest publishing job
5. Copy the entire Tale JSON into the PUT /tale field, move the latest `publishInfo` item to somewhere else in the array, and submit
    * You should see Girder return a 200
6. Refresh the Run > Metadata view
    * You should see that the "Published Location" has not changed, and still displays the same value
